### PR TITLE
Cancel queued tasks

### PR DESCRIPTION
**Pull request by Yadunund**
This PR enables cancellation of of a **queued** task. Support for cancelling an active task is left for future work.

Details of the implementation:
When a `DispatchRequest` message with `CANCEL` method is received by the fleet adapter, it simply re-plans the assignments of all the queued tasks while ignoring the task to be cancelled. The new set of assignments thus will not contain the task to be cancelled. The new assignments are then written to the `TaskManager` queues thereby effectively "cancelling" the requested task. Along the way several checks are included to ensure that the task to be cancelled, actually exists in a queue and is not active.

Lastly this PR introduces a `DispatchAck` message which is published by the fleet adapter in response to a received `DispatchRequest` message. The new message conveys whether the addition or cancellation of a task was successful.

TODO: 
* ~~Update the Dispatcher node to publish a new `BidNotice` only after receiving the `DispatchAck` for the previous bid.~~
* ~~Update the Dispatcher node to cull cancelled tasks from its task registry.~~

### CODE REVIEW

File: rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp [hunk 1]
```diff
@@ -58,37 +58,89 @@ Client::Client(std::shared_ptr<rclcpp::Node> node)
...
-        if (!task_status->is_terminated())
-          _active_task_status[task_id] = task_status;
+      switch (msg->dispatch_request.method)
+      {
+        case RequestMsg::ADD:
+          if (msg->success)
+          {
+            // update this as pending
+            RCLCPP_INFO(_node->get_logger(),
+            "Receive dispatch ack from fleet [%s] that task [%s] is queued",
```
Comment (mxgrey) [inline rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp:93]:
```suggestion
            "Received dispatch ack from fleet [%s] that task [%s] is queued",
```

File: rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp [hunk 2]
```diff
@@ -58,37 +58,89 @@ Client::Client(std::shared_ptr<rclcpp::Node> node)
...
+            msg->dispatch_request.fleet_name.c_str(), task_id.c_str());
+            weak_status->state = TaskStatus::State::Queued;
+          }
+          else
+          {
+            // update this as failed
+            RCLCPP_ERROR(_node->get_logger(),
+            "Receive dispatch ack from fleet [%s] that task [%s] Add Failed",
```
Comment (mxgrey) [inline rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp:101]:
```suggestion
            "Received dispatch ack from fleet [%s] that task [%s] Add failed",
```

File: rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp [hunk 3]
```diff
@@ -58,37 +58,89 @@ Client::Client(std::shared_ptr<rclcpp::Node> node)
...
+            msg->dispatch_request.fleet_name.c_str(), task_id.c_str());
+            weak_status->state = TaskStatus::State::Failed;
+          }
+          break;
+        case RequestMsg::CANCEL:
+          // update this as Canceled
+          if (msg->success)
+          {
+            RCLCPP_INFO(_node->get_logger(),
+            "Receive dispatch ack from fleet [%s] that task [%s] is canceled",
```
Comment (mxgrey) [inline rmf_task_ros2/src/rmf_task_ros2/action/Client.cpp:111]:
```suggestion
            "Received dispatch ack from fleet [%s] that task [%s] is canceled",
```

**Review by mxgrey — APPROVED** [review]
I just left a few minor grammatical suggestions; nothing substantial.

However, I'm seeing that a lot of implementation logic building up in the `rmf_task_ros2` library. Ideally all of that logic should be taken care of in middleware-agnostic `rmf_task` classes with extensive unit testing. The `rmf_task_ros2` classes should only exist to wrap those middleware-agnostic libraries into a ROS2 API. That's a considerable effort so I won't block this feature on that request, but let's try to find time to refactor this implementation.

**Comment by Yadunund** [comment]
Thank you for the review @mxgrey! I believe @tanyouliang95 is working on refactoring `rmf-task_ros2` so as to move a lot of the logic into `rmf_task` in a separate PR as recommended.

**Pull request merged**